service/remote.go

```diff
@@ -140,24 +141,46 @@ func (r *RemoteService) AddRemoteBindingListener(bindingListener cluster.RemoteB
 func (r *RemoteService) Call(ctx context.Context, req *protos.Request) (*protos.Response, error) {
 	c, err := util.GetContextFromRequest(req, r.server.ID)
 	c = util.StartSpanFromRequest(c, r.server.ID, req.GetMsg().GetRoute())
+	defer tracing.FinishSpan(c, err)
 	var res *protos.Response
+
+	if err == nil {
+		result := make(chan *protos.Response, 1)
+		go func() {
+			result <- processRemoteMessage(c, req, r)
+		}()
+
+		// Set a timeout for processing the call
+		timeout := time.Duration(300) * time.Second
+		if ctx.Value("reqTimeout") != nil {
+			timeout, err = time.ParseDuration(fmt.Sprint(ctx.Value("reqTimeout")))
+			if err != nil {
+				logger.Log.Errorf("Error while parsing timeout duration: %s", err.Error())
+			}
+		}
+
+		select {
+		case <-time.After(timeout):
+			err = fmt.Errorf("Timed out calling route %s after %d seconds.", req.GetMsg().GetRoute(), timeout)
+		case res := <-result:
+			return res, nil
+		}
+	}
+
 	if err != nil {
 		res = &protos.Response{
 			Error: &protos.Error{
 				Code: e.ErrInternalCode,
 				Msg:  err.Error(),
 			},
 		}
-	} else {
-		res = processRemoteMessage(c, req, r)
 	}
 
 	if res.Error != nil {
 		err = errors.New(res.Error.Msg)
 	}
 
-	defer tracing.FinishSpan(c, err)
-	return res, nil
+	return res, err
```

Comment:
reminder: this error was always ignored, especially if the error is timeout related, it's important that at least the server logs it. this is also being discarded in the caller function and just passes the error to the response sent to the caller, but the caller might not get that message (it might have already gone away or timed out waiting)

Reply:
addressed